packages/arb-validator/rollupmanager/manager.go

```diff
@@ -79,151 +83,107 @@ func CreateManagerAdvanced(
 	}
+	man.validCallLock.Lock()
 	go func() {
 		for {
 			runCtx, cancelFunc := context.WithCancel(ctx)
 
-			checkpointer := man.ckpFac.New(runCtx)
-
-			var chain *rollup.ChainObserver
+			checkpointer := ckpFac.New(runCtx)
 
 			watcher, err := clnt.NewRollupWatcher(rollupAddr)
 			if err != nil {
 				log.Fatal(err)
 			}
 
-			ethbridgeVersion, err := watcher.GetVersion(runCtx)
-			if err != nil {
-				log.Fatal(err)
-			}
-
-			if ethbridgeVersion != ValidEthBridgeVersion {
-				log.Fatalf("VM has EthBridge version %v, but validator implements version %v."+
-					" To find a validator version which supports your EthBridge, visit "+
-					"https://offchainlabs.com/ethbridge-version-support",
-					ethbridgeVersion, ValidEthBridgeVersion)
-			}
-
-			blockId, initialVMHash, err := watcher.GetCreationInfo(runCtx)
-			if err != nil {
-				log.Fatal(err)
-			}
-
-			initialMachine, err := checkpointer.GetInitialMachine()
+			chain, err := initializeChainObserver(
+				runCtx,
+				rollupAddr,
+				updateOpinion,
+				clnt,
+				watcher,
+				checkpointer,
+			)
 			if err != nil {
 				log.Fatal(err)
 			}
 
-			if initialMachine.Hash() != initialVMHash {
-				log.Fatal("ArbChain was initialized with different VM")
-			}
-
-			if checkpointer.HasCheckpointedState() {
-				err := checkpointer.RestoreLatestState(runCtx, clnt, func(chainObserverBytes []byte, restoreCtx ckptcontext.RestoreContext) error {
-					chainObserverBuf := &rollup.ChainObserverBuf{}
-					if err := proto.Unmarshal(chainObserverBytes, chainObserverBuf); err != nil {
-						log.Fatal(err)
-					}
-					var err error
-					chain, err = chainObserverBuf.UnmarshalFromCheckpoint(runCtx, restoreCtx, checkpointer)
-					return err
-				})
-				if err != nil {
-					log.Fatal(err)
-				}
-			} else {
-				params, err := watcher.GetParams(ctx)
-				if err != nil {
-					log.Fatal(err)
-				}
-				chain, err = rollup.NewChain(rollupAddr, checkpointer, params, updateOpinion, blockId)
-				if err != nil {
-					log.Fatal(err)
-				}
-			}
-
-			log.Println("Starting validator from", chain.CurrentBlockId())
-
-			man.Lock()
-			// Clear pending listeners
-			for len(man.listenerAddChan) > 0 {
-				<-man.listenerAddChan
-			}
+			man.listenersLock.Lock()
+			man.activeChain = chain
+			man.activeCheckpointer = checkpointer
 			// Add manager's listeners
 			for _, listener := range man.listeners {
-				chain.AddListener(listener)
+				man.activeChain.AddListener(listener)
 			}
-			man.Unlock()
+			man.listenersLock.Unlock()
+
+			log.Println("Starting validator from", man.activeChain.CurrentBlockId())
+
+			man.activeChain.RestartFromLatestValid(runCtx)
 
-			chain.Start(runCtx)
+			man.activeChain.Start(runCtx)
 
 			current, err := clnt.CurrentBlockId(runCtx)
 			if err != nil {
 				log.Fatal(err)
 			}
 
-			headersChan, err := clnt.SubscribeBlockHeaders(runCtx, chain.CurrentBlockId())
+			headersChan, err := clnt.SubscribeBlockHeaders(runCtx, man.activeChain.CurrentBlockId())
```

Comment:
Wondering if there is a race condition here. We called RestartFromLatestValid and then Start on man.activeChain above. Now we get CurrentBlockId() from activeChain. Is there a chance that CurrentBlockId changed in the meantime, in a way that causes us to miss some events?

Reply:
And we use the variable "current" in this code as well, as what appears to be a separate notion of what the current block is.

Reply:
If these are indeed different, recommend renaming and/or commenting for clarity.

Reply:
Code has been refactored and variable names changed to increase understandability